Prepare nextcloud storage docs for new authentication method

NobodysNightmare opened the pull request:
This is a technical pre-write of the docs. I've decided to split the instructions per authentication method, because the SSO-specific instructions need to go into a lot of detail in configuring third-party systems. They will also be irrelevant for a large amount of users, because SSO will be a corporate feature.

Proof reading of this draft will certainly be necessary.

# What to consider

This documentation is intended for the ~~15.5~~ 16.0 release. Therefore it's targeting the current `dev` branch and is probably not (yet!) urgent.

# Tickets
* https://community.openproject.org/wp/61652
* https://community.openproject.org/wp/60614

NobodysNightmare committed:
Prepare nextcloud storage docs for new authentication method

This is a technical pre-write of the docs. I've decided to
split the instructions per authentication method, because the
SSO-specific instructions need to go into a lot of detail in
configuring third-party systems. They will also be irrelevant for
a large amount of users, because SSO will be a corporate feature.

Proof reading of this draft will certainly be necessary.

NobodysNightmare committed:
Explain how to migrate to SSO authentication

This covers at least the basic steps necessary.

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
-> These generated values are not accessible again after you close the window. Please do not navigate away from this page before copying them over to Nextcloud, as instructed below. Treat these values with care, as you would an important password. Please do not reveal them to anyone else.
-
-![OpenProject generates OAuth values to copy over to Nextcloud](openproject_system_guide_nextcloud_integration_setup_step_3.png)
+After deciding for an authentication method (more details in the next step), click on **Save and continue setup**. Your new storage is now created, but before you can use it, you will need to configure the chosen authentication method.
```
```suggestion
After deciding on an authentication method (more details in the next step), click on **Save and continue setup**. Your new storage is now created, but before you can use it, you will need to configure the chosen authentication method.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
+### 3. Configure Authentication method
 
-Copy the two generated values (client ID and secret) from the OpenProject tab to the respective fields in Nextcloud, namely **OpenProject OAuth client ID** and **OpenProject OAuth client secret**.
+The next part of the setup will be different depending on the way that you choose authentication between OpenProject and Nextcloud to happen.
```
```suggestion
The next part of the setup will vary depending on your chosen authentication method between OpenProject and Nextcloud.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
+**Two-way OAuth 2.0 authorization code flow:** This way of authentication works with all deployments of OpenProject and Nextcloud. Requests between the two applications will act with the target application acting as OAuth 2.0 authorization server. For the users this means, that they will usually have to confirm a separate OAuth consent screen for both directions of communication.
 
-Once you have copied the values, click on **Save** to proceed to the next step.
+**Single-Sign-On through OpenID Connect Identity Provider:** In this authentication mode, users will not have to confirm additional consent screens, providing a better user experience. Requests between OpenProject and Nextcloud are authenticated with access tokens obtained from a common Identity Provider. This works only if users authenticate to OpenProject and Nextcloud through the same OpenID C ...
```
```suggestion
**Single-Sign-On through OpenID Connect Identity Provider:** This authentication mode improves user experience by eliminating additional consent screens. Requests between OpenProject and Nextcloud are authenticated with access tokens obtained from a common Identity Provider. This requires both applications to use the same OpenID Connect Identity Provider and specific configurations at the provider level. This advanced feature is available for all installations under the Enterprise Corporate plan.```

NobodysNightmare replied:
In general I like this proposal (adding it to the change batch). However, I am not sure how I feel about changing "at the identity provider" to "at the provider level".

"Provider" is a pretty generic term by itself, so keeping it clear with "identity provider" felt better to me.

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -155,6 +132,8 @@ In case you want to use this functionality you will be requested to enter a pass
 > [!IMPORTANT]
 > You will need to install the [Group folder](https://apps.nextcloud.com/apps/groupfolders) app in Nextcloud in order to have OpenProject automatically managed your Nextcloud folders. Each storage can only have one group folder with the same name.
 
+Please note, when you use the **Terms of Service** app on the Nextcloud side, all terms also need to be accepted for the OpenProject user that gets created during the setup. This is set to happen automatically during the initial set-up. If you see an error message indicating otherwise or the integration does not behave as expected, please refer to the [Troubleshooting](#troubleshooting) section a ...
```
```suggestion
> [!NOTE]
> When you use the **Terms of Service** app on the Nextcloud side, all terms also need to be accepted for the OpenProject user that gets created during the setup. This is set to happen automatically during the initial set-up. If you see an error message indicating otherwise or the integration does not behave as expected, please refer to the [Troubleshooting](#troubleshooting) section at the bottom of this page.
```

NobodysNightmare replied:
Sounds good to me. Just the note from my side, that this block was taken as-is from another part of the original docs. So beforehand it wasn't in a "NOTE" box.

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+2. The identity provider needs to hand out compatible access tokens
+3. The identity provider needs to allow long-term access (e.g. through the `offline_access` scope)
+
+Those requirements are explained in more detail below.
```
```suggestion
These requirements are explained in more detail below.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+### Token compatibility
+
+The identity provider will need to hand out JWT encoded access tokens that can be validated according to the rules of [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068). Such tokens indicate at which application they are allowed to be used (the so-called Token _Audience_) and can be validated by an application mostly without performing additional web requests to the identity provider.
```
```suggestion
The identity provider will need to hand out JWT encoded access tokens that can be validated according to the rules of [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068). Such tokens indicate which application they are allowed to be used  in (the so-called Token _Audience_) and can be validated by an application mostly without performing additional web requests to the identity provider.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+1. It allows to perform _OAuth 2.0 Token Exchange_ according to [RFC 8693](https://www.rfc-editor.org/rfc/rfc8693)
+2. Alternatively, it provides tokens with the necessary audience during initial SSO authentication of the user
+
+Token exchange is generally the more secure option, since it allows fine-grained control by the identity provider which application can obtain tokens to make requests to other applications. It's usually harder to use leaked access tokens of this kind for privilege escalation. However, depending on the provider, the latter option may be easier to set-up.
```
```suggestion
Token exchange is generally the more secure option, since it allows stricter control by the identity provider which application can obtain tokens to make requests to other applications. It's usually harder to use leaked access tokens of this kind for privilege escalation. However, depending on the provider, the latter option may be easier to set-up.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+### Keycloak
+
+Keycloak can provide tokens via Token Exchange, as well as immediately after signing in the user. Using Token Exchange is preferred, but requires to enable two preview features in the Keycloak deployment:
```
```suggestion
Keycloak can provide tokens via Token Exchange, as well as immediately after user sign in. Using Token Exchange is preferred, but requires to enable two preview features in the Keycloak deployment:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+* `token-exchange`
+* `admin-fine-grained-authz`
+
+If you prefer to not enable preview features on your production deployment, it's also possible to use wide-access tokens.
```
```suggestion
If you prefer not to enable preview features on your production deployment, it's also possible to use wide-access tokens.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+If you prefer to not enable preview features on your production deployment, it's also possible to use wide-access tokens.
+
+#### ... using Token Exchange
```
Is this heading correct? It looks like the first part might be missing

NobodysNightmare replied:
Feel free to edit. It makes sense to me, when looking at the headings we have:

* Keycloak
    * ... using Token Exchange
    * ... using wide access tokens
* Nextcloud Hub

So the two "..." mean "Keycloak", but I wanted to avoid the repetition.

NobodysNightmare replied:
Dominic also commented this, the headings are now prefixed with "Keycloak" instead of "..."

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+##### Configuring Keycloak
+
+First make sure that you enabled both required features in Keycloak: `token-exchange` and `admin-fine-grained-authz`. You can [follow the Keycloak docs on how to do that](https://www.keycloak.org/server/features). Once they are enabled you can confirm them from the "Server Info" tab in your Keycloak installation:
```
```suggestion
First make sure that you enabled both required features in Keycloak: `token-exchange` and `admin-fine-grained-authz`. You can [follow the Keycloak documentation on how to do that](https://www.keycloak.org/server/features). Once they are enabled you can confirm them from the "Server Info" tab in your Keycloak installation:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+##### Configuring Nextcloud
+
+TODO: How should we configure Nextcloud in this case? Not possible yet? Use `openproject`?
```
Is it now cleared up?

NobodysNightmare replied:
Not yet. Worst-case is that we have to throw it out. We should hopefully have a decision this week.

Background: It was "not yet" possible, but work for that is underway, but requires changes from the Nextcloud developers as well.

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+### Nextcloud Hub
+
+TODO: Maybe a few introductory words about Nextcloud Hub? (e.g. "Nextcloud has two roles; 1. IDP, 2. storage)
```
That would be good

NobodysNightmare replied:
added

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+##### Configuring Nextcloud
+
+TODO: How do we configure this scenario?
```
Is there an update here?

NobodysNightmare replied:
now there is

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+---
+sidebar_navigation:
+  title: Setup Single-Sign-On through OpenID Connect Identity Provider
+  priority: 600
```
```suggestion
  priority: 900
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+---
+sidebar_navigation:
+  title: Setup Two-way OAuth 2.0 authentication
+  priority: 600
```
```suggestion
  priority: 800
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+# Setup Two-way OAuth 2.0 authentication
+
+This authentication mode will use access tokens provided by an OAuth application of either side of the integration between OpenProject and Nextcloud to authenticate requests performed in the name of users.
```
```suggestion
This authentication mode will use access tokens provided by an OAuth application on either side of the integration between OpenProject and Nextcloud to authenticate requests performed in the name of users.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+## 2. Enter Nextcloud OAuth values in OpenProject
+
+In the page that appears, you will see new OAuth values that are once again generated automatically, but this time by Nextcloud.
```
```suggestion
On the page that appears, you will see new OAuth values that are once again generated automatically, but this time by Nextcloud.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+In the page that appears, you will see new OAuth values that are once again generated automatically, but this time by Nextcloud.
+
+OpenProject will need these values to be able to connect to your Nextcloud instance.
```
```suggestion
OpenProject will need these values to be able to connect with your Nextcloud instance.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+![Nextcloud also generates OAuth values that need to be copied to OpenProject](two_way_oauth2_setup_step_3.png)
+
+Navigate back to your OpenProject tab and click on the **Done, continue** button on the screen you previously left it at. You will now see a screen where you will be able to enter the Nextcloud values.
```
```suggestion
Navigate back to your OpenProject tab and click the **Done, continue** button on the screen you previously left it at. You will now see a screen where you will be able to enter the Nextcloud values.
```

MayaBerd reviewed (commented):
I'm done with the first round of reviewing, added some text change suggestions. It looks like there might be an update needed on on the TODO parts of text, @NobodysNightmare.

NobodysNightmare committed:
Apply suggestions from code review

Co-authored-by: Maya Berdygylyjova <MayaBerd@users.noreply.github.com>

NobodysNightmare commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+![OpenProject is configured to manually exchange tokens for the audience called nextcloud](openproject_manual_storage_audience.png)
+
+##### Configuring Nextcloud
```
TODO: Explain how to configure the `user_oidc` app. At least:

1. Enable accepting Bearer Tokens
2. Enable storing login tokens

dominic-braeunlein committed:
docs there -> their fix

dominic-braeunlein commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
+**Two-way OAuth 2.0 authorization code flow:** This way of authentication works with all deployments of OpenProject and Nextcloud. Requests between the two applications will act with the target application acting as OAuth 2.0 authorization server. For the users this means, that they will usually have to confirm a separate OAuth consent screen for both directions of communication.
 
-Once you have copied the values, click on **Save** to proceed to the next step.
+**Single-Sign-On through OpenID Connect Identity Provider:** This authentication mode improves user experience by eliminating additional consent screens. Requests between OpenProject and Nextcloud are authenticated with access tokens obtained from a common Identity Provider. This requires both applications to use the same OpenID Connect Identity Provider and specific configurations at the provide ...
```
I think these ``` are extra
```suggestion
**Single-Sign-On through OpenID Connect Identity Provider:** This authentication mode improves user experience by eliminating additional consent screens. Requests between OpenProject and Nextcloud are authenticated with access tokens obtained from a common Identity Provider. This requires both applications to use the same OpenID Connect Identity Provider and specific configurations at the provider level. This advanced feature is available for all installations under the Enterprise Corporate plan.
```

dominic-braeunlein commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+If you prefer not to enable preview features on your production deployment, it's also possible to use wide-access tokens.
+
+#### ... using Token Exchange
```
I would repeat it. The "using wide access tokens" part is so far down that the context was lost to me
```suggestion
#### Keycloak using Token Exchange
```

dominic-braeunlein commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,167 @@
+
+This wraps up the configuration of this authentication method using Keycloak and token exchange. For details on the next step continue with the [general setup instructions](../#4-automatically-managed-project-folders).
+
+#### ... using wide-access tokens
```
```suggestion
#### Keycloak using wide-access tokens
```

NobodysNightmare committed:
Address further review comments

NobodysNightmare committed:
Update SSO storage docs

Fixing the existing TODOs in the docs and
adding instructions on how to configure scopes correctly.

NobodysNightmare commented:
@MayaBerd I completed the docs (removed remaining TODOs). Everything should be up-to-date right now.

Not sure how the process should optimally go on from your point-of-view. I could use this being merged soon, because we want to link from the application to the docs and there are consistency checks in CI that ensure links are valid. So having some docs would be required for CI to pass once we add those links.

MayaBerd commented:
> @MayaBerd I completed the docs (removed remaining TODOs). Everything should be up-to-date right now.
> 
> Not sure how the process should optimally go on from your point-of-view. I could use this being merged soon, because we want to link from the application to the docs and there are consistency checks in CI that ensure links are valid. So having some docs would be required for CI to pass once we add those links.

Thanks @NobodysNightmare! I'll review by tomorrow morning, so you can make last tweaks if necessary and merge.

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -43,8 +43,8 @@ features and prompt fixes for any potential bugs.
-  - Nextcloud company advises the use of there [all-in-one](https://hub.docker.com/r/nextcloud/all-in-one) docker image.
-- If you want to use the feature of [automatically managed project folders](#5-automatically-managed-project-folders)
+  - Nextcloud company advises the use of their [all-in-one](https://hub.docker.com/r/nextcloud/all-in-one) docker image.
+- If you want to use the feature of [automatically managed project folders](#4-automatically-managed-project-folders)
```
I'd reformulate like this
"To use the [automatically managed project folders feature](#4-automatically-managed-project-folders), you need to install version `14.0.9` or later of the [Group folders app in Nextcloud](https://apps.nextcloud.com/apps/groupfolders), which is the minimum officially supported version."

NobodysNightmare replied:
I can look into this, though note that this is out of scope of my actual changes.

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -81,13 +81,9 @@ Click on the **Save** button.
-### 2. Create a Nextcloud file storage in your OpenProject instance
-
-Navigate to your administration settings page by clicking on *your avatar in the top right corner → Administration*. From the side menu on the left, click on **Files -> External file storages**.
+Navigate to your OpenProject administration settings page by clicking on *your avatar in the top right corner → Administration*. From the side menu on the left, click on **Files -> External file storages**.
```
```suggestion
Navigate to your OpenProject administration settings page by clicking on *your avatar in the top right corner → Administration*. From the side menu on the left select *Files -> External file storages*.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
+After deciding on an authentication method (more details in the next step), click on **Save and continue setup**. Your new storage is now created, but before you can use it, you will need to configure the chosen authentication method.
 
-Go back to the browser tab where you were configuring the **OpenProject Integration** app. (We recommend you have two browser tabs open: the current one with OpenProject and the former one with Nextcloud).
+### 3. Configure Authentication method
```
```suggestion
### 3. Configure authentication method
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
+The next part of the setup will vary depending on your chosen authentication method between OpenProject and Nextcloud.
 
-![OAuth values generated by OpenProject are entered into Nextcloud app configuration](openproject_system_guide_nextcloud_integration_setup_step_4.png)
+**Two-way OAuth 2.0 authorization code flow:** This way of authentication works with all deployments of OpenProject and Nextcloud. Requests between the two applications will act with the target application acting as OAuth 2.0 authorization server. For the users this means, that they will usually have to confirm a separate OAuth consent screen for both directions of communication.
```
```suggestion
**Two-way OAuth 2.0 authorization code flow:** This way of authentication works with all deployments of OpenProject and Nextcloud. Requests between the two applications will act with the target application acting as OAuth 2.0 authorization server. For the users this means that they will usually have to confirm a separate OAuth consent screen for both directions of communication.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
+**Two-way OAuth 2.0 authorization code flow:** This way of authentication works with all deployments of OpenProject and Nextcloud. Requests between the two applications will act with the target application acting as OAuth 2.0 authorization server. For the users this means, that they will usually have to confirm a separate OAuth consent screen for both directions of communication.
 
-Once you have copied the values, click on **Save** to proceed to the next step.
+**Single-Sign-On through OpenID Connect Identity Provider:** This authentication mode improves user experience by eliminating additional consent screens. Requests between OpenProject and Nextcloud are authenticated with access tokens obtained from a common Identity Provider. This requires both applications to use the same OpenID Connect Identity Provider and specific configurations at the provide ...
```
```suggestion
**Single Sign-On through OpenID Connect Identity Provider:** This authentication mode improves user experience by eliminating additional consent screens. Requests between OpenProject and Nextcloud are authenticated with access tokens obtained from a common Identity Provider. This requires both applications to use the same OpenID Connect Identity Provider and specific configurations at the provider level. This advanced feature is available for all installations under the Enterprise Corporate plan.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -103,44 +99,25 @@ Next, enter the **Host URL** of your Nextcloud instance. This is simply the addr
 
-OpenProject will need these values to be able to connect to your Nextcloud instance.
+* [Setup Two-way OAuth 2.0 authorization code flow](./two-way-oauth2/)
+* [Setup Single-Sign-On through OpenID Connect Identity Provider](./oidc-sso/)
```
```suggestion
* [Setup Single Sign-On through OpenID Connect Identity Provider](./oidc-sso/)
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+---
+sidebar_navigation:
+  title: Setup Single-Sign-On through OpenID Connect Identity Provider
```
```suggestion
  title: Setup Single Sign-On through OpenID Connect Identity Provider
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+sidebar_navigation:
+  title: Setup Single-Sign-On through OpenID Connect Identity Provider
+  priority: 900
+description: Set up Single-Sign-On through OpenID Connect Identity Provider as the authentication method for your Nextcloud storage
```
```suggestion
description: Set up Single Sign-On through OpenID Connect Identity Provider as the authentication method for your Nextcloud storage
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+keywords: Nextcloud file storage integration
+---
+
+# Setup Single-Sign-On through OpenID Connect Identity Provider
```
```suggestion
# Setup Single Sign-On through OpenID Connect Identity Provider
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+### User authentication
+
+The requests between OpenProject and Nextcloud usually happen on behalf of a specific user. For example when OpenProject requests a file, Nextcloud will need to make sure that the user for whom we download the file had access to it in the first place.
```
```suggestion
The requests between OpenProject and Nextcloud usually happen on behalf of a specific user. For example, when OpenProject requests a file, Nextcloud will need to make sure that the user for whom we download the file has access to it in the first place.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+For that to be possible, the user must authenticate in both applications through the same identity provider.
+
+Additionally the user account has to be created in both systems already at the time of the request. Right now this requires users to login to both systems at least once. In the future, the initial creations of accounts will also be automated.
```
```suggestion
Additionally, the user account has to be created in both systems already at the time of the request. Right now this requires users to log in to both systems at least once. In the future, the initial creation of accounts will also be automated.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+### Token compatibility
+
+The identity provider will need to hand out JWT encoded access tokens that can be validated according to the rules of [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068). Such tokens indicate which application they are allowed to be used  in (the so-called Token _Audience_) and can be validated by an application mostly without performing additional web requests to the identity provider.
```
```suggestion
The identity provider will need to hand out JWT-encoded access tokens that can be validated according to the rules of [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068). Such tokens indicate which application they are allowed to be used  in (the so-called Token _Audience_) and can be validated by an application mostly without performing additional web requests to the identity provider.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+The identity provider will need to hand out JWT encoded access tokens that can be validated according to the rules of [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068). Such tokens indicate which application they are allowed to be used  in (the so-called Token _Audience_) and can be validated by an application mostly without performing additional web requests to the identity provider.
+
+It must be possible for OpenProject to obtain an access token that includes the Nexcloud client in the audience and vice versa, to allow making requests between the two applications. This has to be supported in one of two ways by the identity provider:
```
```suggestion
It must be possible for OpenProject to obtain an access token that includes the Nextcloud client in the audience and vice versa, to allow making requests between the two applications. This has to be supported in one of two ways by the identity provider:
```

MayaBerd replied:
```suggestion
It must be possible for OpenProject to obtain an access token that includes the Nextcloud client in the audience and vice versa, to allow making requests between the two applications. This has to be supported in one of two ways by the identity provider:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+It must be possible for OpenProject to obtain an access token that includes the Nexcloud client in the audience and vice versa, to allow making requests between the two applications. This has to be supported in one of two ways by the identity provider:
+
+1. It allows to perform _OAuth 2.0 Token Exchange_ according to [RFC 8693](https://www.rfc-editor.org/rfc/rfc8693)
```
```suggestion
1. It allows performing _OAuth 2.0 Token Exchange_ according to [RFC 8693](https://www.rfc-editor.org/rfc/rfc8693)
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+1. It allows to perform _OAuth 2.0 Token Exchange_ according to [RFC 8693](https://www.rfc-editor.org/rfc/rfc8693)
+2. Alternatively, it provides tokens with the necessary audience during initial SSO authentication of the user
+
+Token exchange is generally the more secure option, since it allows stricter control by the identity provider which application can obtain tokens to make requests to other applications. It's usually harder to use leaked access tokens of this kind for privilege escalation. However, depending on the provider, the latter option may be easier to set-up.
```
```suggestion
Token exchange is generally the more secure option, since it allows stricter control by the identity provider which application can obtain tokens to make requests to other applications. It's usually harder to use leaked access tokens of this kind for privilege escalation. However, depending on the provider, the latter option may be easier to set up.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+### Long-term access
+
+OpenProject relies on the ability to retain access to issued tokens over a long duration. That means issued refresh tokens should not expire under normal circumstances. This is usually achieved by requesting a scope called `offline_access` from the identity provider.
```
```suggestion
OpenProject relies on the ability to retain access to issued tokens over a long duration. This means that issued refresh tokens should not expire under normal circumstances. This is usually achieved by requesting the `offline_access` scope from the identity provider.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+## 2. Configuring supported identity providers
+
+The integration between OpenProject and Nextcloud has been tested with the following identity providers and we consider them officially supported:
```
```suggestion
The integration between OpenProject and Nextcloud has been tested with the following identity providers, which we consider officially supported:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+* Keycloak
+* Nextcloud Hub
+
+Other identity providers might work as well, but we only have limited capabilities to help you in setting them up. The Keycloak instructions should explain all the general steps necessary to make it work, but the exact way to configure it for your identity provider will most likely be different.
```
```suggestion
Other identity providers might work as well, but we have only limited capacity to assist with their setup. The Keycloak instructions should explain all the general steps necessary to make it work, but the exact way to configure it for your identity provider will likely differ.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+### Keycloak
+
+Keycloak can provide tokens via Token Exchange, as well as immediately after user sign in. Using Token Exchange is preferred, but requires to enable two preview features in the Keycloak deployment:
```
```suggestion
Keycloak can provide tokens via Token Exchange, as well as immediately after user sign-in. Using Token Exchange is preferred, but requires enabling two preview features in the Keycloak deployment:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+#### Common setup steps
+
+Regardless of the way that tokens are handed out, you need to configure the clients correctly. Assuming that users can already sign in
```
```suggestion
Regardless of how tokens are handed out, you need to configure the clients correctly. Assuming that users can already sign in
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+#### Common setup steps
+
+Regardless of the way that tokens are handed out, you need to configure the clients correctly. Assuming that users can already sign in
+to your OpenProject and Nextcloud instance via Keycloak, you should already have configured a client for both applications. For the rest of
```
```suggestion
to your OpenProject and Nextcloud instances via Keycloak, you should already have configured a client for both applications. For the rest of
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+Regardless of the way that tokens are handed out, you need to configure the clients correctly. Assuming that users can already sign in
+to your OpenProject and Nextcloud instance via Keycloak, you should already have configured a client for both applications. For the rest of
+this guide we will assume that these clients exist with the client ID `openproject` and `nextcloud` respectively.
```
```suggestion
this guide, we will assume that these clients exist with the client ID `openproject` and `nextcloud` respectively.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+to your OpenProject and Nextcloud instance via Keycloak, you should already have configured a client for both applications. For the rest of
+this guide we will assume that these clients exist with the client ID `openproject` and `nextcloud` respectively.
+
+To make requests to the API of OpenProject a client needs to obtain a token with the scope `api_v3`. We need to configure this client scope
```
```suggestion
To make requests to the API of OpenProject, a client needs to obtain a token with the scope `api_v3`. We need to configure this client scope
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+To make requests to the API of OpenProject a client needs to obtain a token with the scope `api_v3`. We need to configure this client scope
+in Keycloak, so that the Nextcloud client can obtain it.
+
+First navigate to "Client scopes" and there press the button to "Create client scope" and give the scope a name of `api_v3`. Make sure
```
```suggestion
First, navigate to "Client scopes", click the "Create client scope" button and name the scope `api_v3`. Make sure
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+First navigate to "Client scopes" and there press the button to "Create client scope" and give the scope a name of `api_v3`. Make sure
+to enable the option "Include in token scope", then save the form.
+
+This makes the scope ready to use. How we will use it, depends on the way that tokens are handed out to clients.
```
```suggestion
This makes the scope ready to use. How we use it depends on the way that tokens are handed out to clients.
```

NobodysNightmare replied:
This doesn't sound right. "How use it" sounds like there is a word missing.

Maybe "how to"?

MayaBerd replied:
Yes, a typo on my side. I've corrected the suggestion.

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+##### Configuring Keycloak
+
+First make sure that you enabled both required features in Keycloak: `token-exchange` and `admin-fine-grained-authz`. You can [follow the Keycloak documentation on how to do that](https://www.keycloak.org/server/features). Once they are enabled you can confirm them from the "Server Info" tab in your Keycloak installation:
```
```suggestion
First, make sure that you enabled both required features in Keycloak: `token-exchange` and `admin-fine-grained-authz`. You can [follow the Keycloak documentation on how to do that](https://www.keycloak.org/server/features). Once they are enabled, you can confirm them from the "Server Info" tab in your Keycloak installation:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+We will need to allow OpenProject to exchange tokens with `nextcloud` in the audience, as well as allowing Nextcloud to exchange tokens with `openproject` in the audience.
+
+To do this, we first create policies that will be used to control which clients are allowed to exchange tokens for a specific audience. Go to the `realm-management` client (in the default realm this client is called `master-realm`) and switch to the Tab "Authorization" and there to the "Policies" submenu. In case the "Authorization" tab is not visible, it might be necessary to make it visible by  ...
```
```suggestion
To do this, we first create policies that will be used to control which clients are allowed to exchange tokens for a specific audience. Go to the `realm-management` client (in the default realm this client is called `master-realm`) and switch to the "Authorization" tab and there to the "Policies" submenu. If the "Authorization" tab is not visible, it might be necessary to make it visible by going to "Permissions" first and enabling fine-grained permissions there. Assuming you didn't create any other policies yet, your view should look similar to this:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+![One of the two exchange policies configured above](keycloak_exchange_policy.png)
+
+We will later use these policies to control who is allowed to perform a token exchange that results in a token usable for requests to the given client. For example if you have additional clients that need to obtain a token usable at OpenProject, you'd add those clients to the `allowed-to-exchange-openproject-token` client list. The names suggested here are proposals for readable policy names, but ...
```
```suggestion
We will later use these policies to control who is allowed to perform a token exchange that results in a token usable for requests to the given client. For example, if you have additional clients that need to obtain a token usable at OpenProject, you'd add those clients to the `allowed-to-exchange-openproject-token` client list. The names suggested here are proposals for readable policy names, but they have no practical impact.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+We will later use these policies to control who is allowed to perform a token exchange that results in a token usable for requests to the given client. For example if you have additional clients that need to obtain a token usable at OpenProject, you'd add those clients to the `allowed-to-exchange-openproject-token` client list. The names suggested here are proposals for readable policy names, but ...
+
+Next we will need to use the policies, so that they actually do allow the token exchange for given clients. Navigate to the `openproject` client, go to the Tab "Permissions", enable the box "Permissions enabled" and select the permission with the name `token-exchange`. You should end up on a form similar to this one:
```
```suggestion
Next, we need to apply the policies, so that they actually allow the token exchange for given clients. Navigate to the `openproject` client, go to the Tab "Permissions", enable the "Permissions enabled" checkbox and select the permission with the name `token-exchange`. You should end up on a form similar to this one:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+![Empty view of the permission configuration dialog](keycloak_exchange_permission.png)
+
+In the field "Policies" you can now add the `allowed-to-exchange-openproject-token` we created earlier. This means that clients matching that policy will be allowed to exchange an access token that has the `openproject`.
```
```suggestion
In the field "Policies" you can now add the `allowed-to-exchange-openproject-token` we created earlier. This means that clients matching that policy will be allowed to exchange an access token that includes `openproject`.
```

NobodysNightmare replied:
My original sentence was incomplete. It should have said "has the `openproject` audience. But I'll do the completed sentence using "includes" as suggested.

NobodysNightmare replied:
done

NobodysNightmare committed:
Apply suggestions from code review

Co-authored-by: Maya Berdygylyjova <MayaBerd@users.noreply.github.com>

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+##### Configuring Nextcloud
+
+In Nextcloud we need to configure that we are using an "external provider" by selecting the corresponding radio button in the authentication settings of the OpenProject integration. Select the OIDC provider that we are using as a common identity provider from the dropdown.
```
```suggestion
In Nextcloud, we need to configure that we are using an "external provider" by selecting the corresponding radio button in the authentication settings of the OpenProject integration. Select the OIDC provider we’re using as the common identity provider from the dropdown.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+![Nextcloud configured to use the login token](nextcloud_using_login_token.png)
+
+Additionally we need to configure the OpenID Connect provider in Nextcloud. Navigate to "Administration", and choose "OpenID Connect" from the sidebar. You should see a list of your configured OpenID Connect providers. At the top of that page enable the checkbox called "Store login tokens". This setting is required by Nextcloud to allow using tokens obtained during login for token exchange or req ...
```
```suggestion
Additionally, we need to configure the OpenID Connect provider in Nextcloud. Navigate to "Administration" and choose "OpenID Connect" from the sidebar. You should see a list of your configured OpenID Connect providers. At the top of that page enable the "Store login tokens" checkbox. This setting is required by Nextcloud to allow using tokens obtained during login for token exchange or requests to other services.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+Additionally we need to configure the OpenID Connect provider in Nextcloud. Navigate to "Administration", and choose "OpenID Connect" from the sidebar. You should see a list of your configured OpenID Connect providers. At the top of that page enable the checkbox called "Store login tokens". This setting is required by Nextcloud to allow using tokens obtained during login for token exchange or req ...
+
+Afterwards navigate to the OpenID Connect provider of the Keycloak and edit it. Make sure that the `api_v3` claim is part of the "Scope" text box. If it hasn't been added so far, you can append it to the existing text, ensuring that it is preceeded by a space, e.g., `openid email profile api_v3`. This ensures that Nextcloud will request the `api_v3` scope when a user logs in. Only then Keycloak w ...
```
```suggestion
Afterwards navigate to the OpenID Connect provider of the Keycloak and edit it. Make sure that the `api_v3` claim is part of the "Scope" text box. If it hasn't been added so far, you can append it to the existing text, ensuring that it is preceeded by a space, e.g., `openid email profile api_v3`. This ensures that Nextcloud will request the `api_v3` scope when a user logs in. Only then will Keycloak allow to exchange a token for OpenProject that carries the `api_v3` scope as well.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+* It acts as a storage provider, just as usual
+* It acts as the identity provider
+
+This deployment is simpler compared to running a dedicated identity provider, because it requires deployment and configuration of fewer applications.
```
```suggestion
This deployment is simpler than running a dedicated identity provider, because it requires deployment and configuration of fewer applications.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+##### Configuring OpenProject
+
+In OpenProject configure the storage to "Use first access token obtained by identity provider". This means OpenProject will not attempt to perform a token exchange, but simply use the first token returned by the identity provider for authentication against the storage as well:
```
```suggestion
In OpenProject, configure the storage to "Use first access token obtained by identity provider". This means OpenProject will not attempt to perform a token exchange, but simply use the first token returned by the identity provider for authentication against the storage as well:
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+
+##### Configuring Nextcloud
+
+In Nextcloud we configure the OIDC provider to be "Nextcloud Hub". In the field "OpenProject client ID" you have to set the client ID of OpenProject as it's configured in Nextcloud (in the sidebar select "Security" and then scroll down to "OpenID Connect clients"):
```
```suggestion
In Nextcloud, we configure the OIDC provider to be "Nextcloud Hub". In the field "OpenProject client ID" you have to set the client ID of OpenProject as it's configured in Nextcloud (in the sidebar select "Security" and then scroll down to "OpenID Connect clients"):
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/oidc-sso/README.md:
```diff
@@ -0,0 +1,211 @@
+It is possible to migrate an existing storage to use SSO authentication. You can follow the steps outlined above for your specific setup, all of the relevant options
+in OpenProject and Nextcloud are also accessible from the views to edit your storage.
+
+Please note that all the [prerequisites](#1-prerequisites) mentioned above still apply, this means if your users don't already authenticate to OpenProject and Nextcloud through the same
```
```suggestion
Please note that all the [prerequisites](#1-prerequisites) mentioned above still apply. This means if your users don't already authenticate to OpenProject and Nextcloud through the same
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+---
+sidebar_navigation:
+  title: Setup Two-way OAuth 2.0 authentication
```
```suggestion
  title: Two-way OAuth 2.0 authentication setup
```

NobodysNightmare replied:
Mh. I am indifferent about how we should write this, but I'd expect this to be consistent across SSO and and Two-way OAuth 2.0.

Right now your suggestion would make it inconsistent across the two. Should I update both to "setup as a suffix" or should I keep both at using "setup as a prefix"?

NobodysNightmare replied:
I am going to merge this as-is now (i.e. with all your other suggestions applied) to unblock PRs that depend on the existence of the docs.

If we need to change this, I can still do it separately.

MayaBerd replied:
@NobodysNightmare I would be ok either way, but if we keep it up front, then i'd rather use the verb, so "Set up", instead of "Setup". 

SSO: yes, good point. I vote for consistency also and would also adjust it on the website and in user guide to "single sign-on".

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+keywords: Nextcloud file storage integration
+---
+
+# Setup Two-way OAuth 2.0 authentication
```
```suggestion
# Two-way OAuth 2.0 authentication setup
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+# Setup Two-way OAuth 2.0 authentication
+
+This authentication mode will use access tokens provided by an OAuth application on either side of the integration between OpenProject and Nextcloud to authenticate requests performed in the name of users.
```
```suggestion
This authentication mode uses access tokens provided by an OAuth application on either side of the integration between OpenProject and Nextcloud to authenticate requests performed in the name of users.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+This authentication mode will use access tokens provided by an OAuth application on either side of the integration between OpenProject and Nextcloud to authenticate requests performed in the name of users.
+
+You will have to perform configuration in OpenProject and in Nextcloud.
```
```suggestion
You will need to perform configuration in OpenProject and in Nextcloud.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+After selecting and saving the authentication method in OpenProject, you will see a form with two text fields that show a **client ID** and a **client secret** that belong to the OAuth application that OpenProject has just automatically created.
+
+These values are needed to permit Nextcloud to connect to OpenProject.
```
```suggestion
These values are required to permit Nextcloud to connect to OpenProject.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+These values are needed to permit Nextcloud to connect to OpenProject.
+
+> [!IMPORTANT]
+> These generated values are not accessible again after you close the window. Please do not navigate away from this page before copying them over to Nextcloud, as instructed below. Treat these values with care, as you would an important password. Please do not reveal them to anyone else.
```
```suggestion
> These generated values are not accessible again after you close the window. Please do not navigate away from this page before copying them over to Nextcloud, as instructed below. Treat these values with care, as you would an important password. Please do not share them with anyone.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+![OpenProject generates OAuth values to copy over to Nextcloud](two_way_oauth2_setup_step_1.png)
+
+Go back to the browser tab where you were configuring the **OpenProject Integration** app. (We recommend you have two browser tabs open: the current one with OpenProject and the former one with Nextcloud).
```
```suggestion
Go back to the browser tab where you were configuring the **OpenProject Integration** app. (We recommend keeping two browser tabs open: one for OpenProject and one for Nextcloud).
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+Go back to the browser tab where you were configuring the **OpenProject Integration** app. (We recommend you have two browser tabs open: the current one with OpenProject and the former one with Nextcloud).
+
+Copy the two generated values (client ID and secret) from the OpenProject tab to the respective fields in Nextcloud, namely **OpenProject OAuth client ID** and **OpenProject OAuth client secret**.
```
```suggestion
Copy the two generated values (client ID and secret) from the OpenProject tab into the corresponding fields in Nextcloud, namely **OpenProject OAuth client ID** and **OpenProject OAuth client secret**.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+## 2. Enter Nextcloud OAuth values in OpenProject
+
+On the page that appears, you will see new OAuth values that are once again generated automatically, but this time by Nextcloud.
```
```suggestion
On the page that appears, you will see new OAuth values that are again generated automatically, this time by Nextcloud.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+OpenProject will need these values to be able to connect with your Nextcloud instance.
+
+Much like in the previous step, you will need to copy these two generated values (**Nextcloud OAuth client ID** and **Nextcloud OAuth client secret**) and paste them into OpenProject.
```
```suggestion
As in the previous step, copy these two generated values (**Nextcloud OAuth client ID** and **Nextcloud OAuth client secret**) and paste them into OpenProject.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+![Nextcloud also generates OAuth values that need to be copied to OpenProject](two_way_oauth2_setup_step_3.png)
+
+Navigate back to your OpenProject tab and click the **Done, continue** button on the screen you previously left it at. You will now see a screen where you will be able to enter the Nextcloud values.
```
```suggestion
Navigate back to your OpenProject tab and click the **Done, continue** button on the screen you previously left it. You will now see a screen where you will be able to enter the Nextcloud values.
```

MayaBerd commented on docs/system-admin-guide/integrations/nextcloud/two-way-oauth2/README.md:
```diff
@@ -0,0 +1,50 @@
+
+Once you have entered the client ID and client secrets on this page, click on **Save and continue**.
+
+This wraps up the configuration of the authentication method. For details on the next step continue with the [general setup instructions](../#4-automatically-managed-project-folders).
```
```suggestion
This wraps up the configuration of the authentication method. For details on the next step, continue with the [general setup instructions](../#4-automatically-managed-project-folders).
```

MayaBerd reviewed (approved):
@NobodysNightmare i've added my review with some suggestions on wording.

NobodysNightmare committed:
Apply suggestions from code review

Co-authored-by: Maya Berdygylyjova <MayaBerd@users.noreply.github.com>
